### ISSUE TITLE
False positive in python.lang.correctness.tempfile.flush.tempfile-without-flush without explicit write calls

### ISSUE THREAD

**Issue by huonw** [returntocorp/sgrep-rules#1448]
**Describe the bug**

It seems to be possible to use the `.name` property of a temporary file without using `.flush` or `.close` causing issues in a way that is still flagged by semgrep.

**To Reproduce**

```python
import tempfile
import subprocess

with tempfile.NamedTemporaryFile() as tmp:
    # tmp.flush()
    # do something like invoke a command that writes to this file
    subprocess.run(f"echo a > {tmp.name}", shell=True)

    # now read the results
    print(tmp.read())
```

```
$ semgrep --config="r/python.lang.correctness.tempfile.flush.tempfile-without-flush" tmp_false_positive.py
using config from https://semgrep.dev/r/python.lang.correctness.tempfile.flush.tempfile-without-flush. Visit https://semgrep.dev/registry to see all public rules.
downloading config...
running 1 rules...
.../tmp_false_positive.py
severity:error rule:python.lang.correctness.tempfile.flush.tempfile-without-flush: Using 'tmp.name' without '.flush()' or '.close()' may cause an error because the file may not exist when 'tmp.name' is used. Use '.flush()' or close the file before using 'tmp.name'.
7:    subprocess.run(f"echo a > {tmp.name}", shell=True)
ran 1 rules on 1 files: 1 findings
```

The same error appears if the `tmp.flush()` line is uncommented, making the error message misleading.

**Expected behavior**

This particular code works, and, I think, always works, and so the given error shouldn't appear. Additionally, I'm not sure I understand the error message: the Python docs are slightly ambiguous but it certainly sounds like the file itself will exist (emphasis mine):

> This function operates exactly as TemporaryFile() does, except that the file is guaranteed to have a visible name in the file system (on Unix, the directory entry is not unlinked). That name can be retrieved from the name attribute of the returned file-like object

(From the way the rule is written, maybe it is trying to flag when data is written but not flushed, and so an external program may not find it.)

That is, it feels like the rule could avoid flagging uses of `.name` if there's been no other uses of the file object before that property access. Alternatively, if a `.flush` is actually required, the error should be resolved just by adding a `flush`.

(As an additional note about the message, calling `.close` will only work to resolve this problem on a `NamedTemporaryFile(delete=False)` instance: in the default case (or, equivalently, with an explicit `delete=True`) `close` will delete the file, so accessing `.name` after that won't work as desired either, only `.flush` is the appropriate fix in that case. There's potentially scope for an additional rule here checking for use of `TemporaryFile`/`NamedTemporaryFile` after a `.close` call.)

**Priority**
How important is this to you?
- [ ] P0: blocking me from making progress
- [ ] P1: this will block me in the near future
- [x] P2: annoying but not blocking me

**Additional Context**

Semgrep version: 0.63.0
Tag of the rule: sha256:806c0768af02e3fb544db27af0ccf39ad0a4b89f3a4818fa85e3bcbca01f8540

**Comment by minusworld** [comment]
@huonw Thanks for the bug report. I'm actually not certain the original intent of this rule aside from what's in the message.

I looked around on the web and found a [comment on this answer](https://stackoverflow.com/a/21944302) that seems to indicate that there are circumstances where calling `.name` without `.flush(...)` could cause a subprocess error. However, that discussion was 7 years ago and may no longer be relevant.

I've also tested it locally (I'm on Mac OS) and am unable to replicate the issue the rule describes, so I'm wondering if this rule is even relevant anymore. The git history says I originally wrote the rule, but I'm pretty sure I didn't because I don't remember anything about it :X I'll see if anyone remembers it, and if not, we may remove the rule altogether.

**Comment by minusworld** [comment]
We had some discussion about this a while back but I forgot to revisit. The consensus is that this rule doesn't appear to be relevant in most cases anymore, so we'll probably remove it. Thanks!
